packages/mui-system/src/createStyled.js

```diff
@@ -57,17 +64,31 @@ const getVariantStyles = (name, theme) => {
 const variantsResolver = (props, styles, variants) => {
   const { ownerState = {} } = props;
   const variantsStyles = [];
+  let numOfCallbacks = 0;
 
   if (variants) {
     variants.forEach((variant) => {
       let isMatch = true;
-      Object.keys(variant.props).forEach((key) => {
-        if (ownerState[key] !== variant.props[key] && props[key] !== variant.props[key]) {
-          isMatch = false;
-        }
-      });
+      if (typeof variant.props === 'function') {
+        const propsToCheck = deepmerge(props, ownerState);
```

Comment:
```suggestion
        const propsToCheck = { ...props, ...ownerState };
```
No need to deep merge, it's too expensive and can cause issues if a prop is a styled component (max stack exceeded).